### DISCUSSION TITLE
FSM-like nested timers

### DISCUSSION THREAD

**Discussion by emilyjanedennis** [bonsai-rx/bonsai#910] [Q&A] [answered]
hi all -  I'm trying to track a mouse and trigger a series of events when it is not moving. I think there's about a billion ways to do this in Bonsai, but I'm struggling to find something robust that handles all of the conditions with my limited knowledge of Rx/Bonsai. Any and all suggestions welcome, thanks!

There are three main conditions:
1. Mouse has been moving and is still moving -> do nothing
2. Mouse has been moving but stops moving -> start a timer -> if the mouse still has not moved, trigger an output (serial string write)
3. Mouse was stopped, we already triggered an output, now *start a second timer -> if the mouse still has not moved, trigger an output, repeat this as long as the mouse is not moving*
4. Mouse was stopped but started moving again -> do nothing

I attempted this by taking the movement (int) and thresholding by Less Than, followed by a DistinctUntilChanged node. This way I know when movement starts or stops. One example of something I tried is to make a SelectMany node that takes the source (bool from DistinctUntilChanged) and if TRUE (means animal stopped moving) I delay, trigger output SerialStringWrite, branch from delay1 and introduce a delay2 and then trigger output SerialStringWrite and then Repeat. I think this should repeat the branch with delay2 and SerialStringWrite.

 In parallel, inside the SelectMany node, I take the same original DistinctUntilChanged boolean, use a condition for false, and then connect to workflow output. This successfully waits until the movement stops, but then repeats forever instead of exiting. I am pretty stuck on how to stop the loop once it starts -- I also tried some gates and throttles along the way, as well as triggered buffer and triggered window, but couldn't get them working well and am not able to find enough examples like this to hack away at it.

Thanks for any/all advice!

**Comment by bruno-f-cruz** [comment, marked as answer]
Heya,
This sounds like a perfect case to use the `Switch` combinator!
Basically, the `Switch` closes the current observable once a new one is created. You can leverage this behavior to alternate between the conditions (i.e. Moving -> not moving and vice-versa).

Check the attached workflow for a possible solution (i think....)
Let me know if it works !

[example_fsm.zip](https://github.com/bonsai-rx/bonsai/files/8778689/example_fsm.zip)

**Reply by emilyjanedennis** [reply]
that is SLICK- thank you! works great.
